tsdb/chunkenc/histogram.go

```diff
@@ -612,6 +617,120 @@ func (a *HistogramAppender) writeSumDelta(v float64) {
 	xorWrite(a.b, v, a.sum, &a.leading, &a.trailing)
 }
 
+func (a *HistogramAppender) AppendOrCreate(int64, float64) (Chunk, Appender, error) {
+	panic("appended a float sample to a histogram chunk")
+}
+
+func (a *HistogramAppender) AppendOrCreateFloatHistogram(*FloatHistogramAppender, int64, *histogram.FloatHistogram, bool) (Chunk, bool, Appender, error) {
+	panic("appended a float histogram sample to a histogram chunk")
+}
+
+func (a *HistogramAppender) AppendOrCreateHistogram(prev *HistogramAppender, t int64, h *histogram.Histogram, appendOnly bool) (Chunk, bool, Appender, error) {
+	if a.NumSamples() == 0 {
+		a.AppendHistogram(t, h)
+		if h.CounterResetHint == histogram.GaugeType {
+			a.setCounterResetHeader(GaugeType)
+			return nil, false, a, nil
+		}
+
+		if prev != nil && h.CounterResetHint != histogram.CounterReset {
+			// This is a new chunk, but continued from a previous one. We need to calculate the reset header unless already set.
+			_, _, _, counterReset := prev.Appendable(h)
+			if counterReset {
+				a.setCounterResetHeader(CounterReset)
+			} else {
+				a.setCounterResetHeader(NotCounterReset)
```

Comment:
is this needed? @beorn7

Reply:
Yes, I think so.
We are currently not trusting this hint because the previous chunk might have been deleted, but we will maybe refine our tracking here. For example, chunks don't get deleted from the middle of a block, so in that case, the `NotCounterReset` hint will be helpful.